CesiumGltfWriter/src/EncodeBase64String.cpp

```diff
@@ -1,6 +1,17 @@
 #include "EncodeBase64String.h"
-#include <base64_encode.hpp>
+#include <modp_b64.h>
+
 std::string
 CesiumGltf::encodeAsBase64String(const std::vector<std::byte>& data) noexcept {
-  return base64_encode(data);
+  const std::size_t len = modp_b64_encode_len(data.size());
+  std::vector<char> destination(len, 0);
+  const char* asCharPointer = reinterpret_cast<const char*>(data.data());
+  modp_b64_encode(destination.data(), asCharPointer, data.size());
+
+  // TODO: Returning std:string(destination.begin(), destination.end())
+  //       adds *ALL* the padding bytes instead of just stopping at the first
+  //       null byte. It seems more correct to do that, however, if I do that
+  //       buffer.cesium.data is initalized to nothing when CesiumGltfReader
+  //       reads the glTF asset back in. Bug?
```

Comment:
`modp_b64_encode_len` returns the _max_ length of the encoded string, not the actual length. You should use the return value of `modp_b64_encode` as the length of the buffer. Actually it returns the length with the null terminator, so you should subtract one.

You can also read directly into a string, rather than reading into a vector and then copying into a string. Then resize the string to the actual encoded length at the end.

Reply:
Changed to remove the extra copy, but I'm not sure that minus one is correct:

```cpp
std::string
CesiumGltf::encodeAsBase64String(const std::vector<std::byte>& data) noexcept {
  const std::size_t paddedLength = modp_b64_encode_len(data.size());
  std::string result(paddedLength, '\0');
  const char* asCharPointer = reinterpret_cast<const char*>(data.data());
  const std::size_t actualLength = modp_b64_encode(result.data(), asCharPointer, data.size());
  result.erase(actualLength, std::string::npos);
  return result;
}
```

If `data` == `HelloWorld!`, then `result` should be set to `SGVsbG9Xb3JsZCE=0`, which is `17` characters [0-16]. `paddedLength` is set to `17` and `actualLength` is set to `16`. The docs for the function say it returns the length of the destination plus one for the null byte. As of C++11 [strings are required to be null terminated internally](https://stackoverflow.com/a/7554172). So resizing `result` to be `actualLength - 1` would result in `SGVsbG9Xb3JsZCE` which removes the '=' and generates an invalid base64 string as it would strip the padding. We can erase from `actualLength` to the end of the string to remove the extra bytes instead. I added extra tests for this.

Reply:
It seems the modp_b64 docs are wrong. The length returned by `modp_b64_encode` does _not_ include the null terminator. Looking at the source, the return value is:

```
    *p = '\0';
    return p - (uint8_t*)dest;
```

`dest` is the originall passed in pointer. So if we wrote one character, then incremented `p` and wrote the null character, then `p-dest` would give us 1, which is the number of characters written _without_ the null character. So yeah, the docs are definitely wrong.

So your code is correct, and `result.resize(actualLength)` would work well, too.